homeassistant/components/squeezebox/__init__.py

```diff
@@ -1 +1,76 @@
-"""The squeezebox component."""
+"""The Logitech Squeezebox integration."""
+
+import asyncio
+import logging
+
+from pysqueezebox import async_discover
+
+from homeassistant.components.media_player import DOMAIN as MP_DOMAIN
+from homeassistant.config_entries import SOURCE_DISCOVERY, ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_PORT, EVENT_HOMEASSISTANT_START
+from homeassistant.core import HomeAssistant
+
+from .const import DOMAIN, PLAYER_DISCOVERY_UNSUBS
+
+_LOGGER = logging.getLogger(__name__)
+
+DISCOVERY_TASK = "discovery_task"
+
+
+async def start_server_discovery(hass):
+    """Start a server discovery task."""
+
+    def _discovered_server(server):
+        asyncio.create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_DISCOVERY},
+                data={
+                    CONF_HOST: server.host,
+                    CONF_PORT: int(server.port),
+                    "uuid": server.uuid,
+                },
+            )
+        )
+
+    hass.data.setdefault(DOMAIN, {})
+    if DISCOVERY_TASK not in hass.data[DOMAIN]:
```

Comment:
There is a `TOCTOU` race condition here since there isn't an `asyncio.Lock` when checking to see if the discovery task is in `hass.data[DOMAIN]`

Reply:
Disregard, I don't think it can actually ever happen.

Reply:
Ah. If I instantiate an `asyncio.Lock` and then prefix the `if DISCOVERY_TASK` block with `async with lock:` that should remove the race condition, right? It's not hypothetical since `start_server_discovery` is getting called from both `async_setup` and `async_setup_entry`.